File: backend/app/api/v1/diaries.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from typing import List
from datetime import datetime
from app.db.base import get_db
from app.db import crud
from app.schemas.diary import DiaryCreate, DiaryResponse

router = APIRouter()

# User endpoints removed - now handled by auth.py

@router.post("/diaries", response_model=DiaryResponse)
async def create_or_update_diary(
    diary: DiaryCreate, 
    user_id: int,  # Required user_id parameter
    db: Session = Depends(get_db)
):
    """Create or update a diary entry for the given date"""
    # Verify user exists
    db_user = crud.get_user(db, user_id=user_id)
    if db_user is None:
        raise HTTPException(status_code=404, detail="User not found")
    
    return crud.upsert_diary(
        db=db, 
        user_id=user_id, 
        content=diary.content,
        emotion=diary.emotion,
        music=diary.music,
        date_str=diary.date
    )

@router.get("/diaries", response_model=List[DiaryResponse])
async def get_diaries(
    user_id: int,  # Required user_id parameter
    limit: int = 30, 
    db: Session = Depends(get_db)
):
    """Get user's diary entries"""
    # Verify user exists
    db_user = crud.get_user(db, user_id=user_id)
    if db_user is None:
        raise HTTPException(status_code=404, detail="User not found")
    
    return crud.get_diaries(db, user_id=user_id, limit=limit)

@router.get("/diaries/{diary_id}", response_model=DiaryResponse)
async def get_diary(diary_id: int, db: Session = Depends(get_db)):
    """Get specific diary entry"""
    db_diary = crud.get_diary(db, diary_id=diary_id)
    if db_diary is None:
        raise HTTPException(status_code=404, detail="Diary not found")
    return db_diary

File: frontend/src/api/diaries.ts
import api from './client'
import { getUserInfo } from './auth'

export interface DiaryCreate {
  content: string
  emotion?: string
  music?: any
  date?: string
}

export interface DiaryResponse {
  id: number
  user_id: number
  date: string
  content: string
  emotion?: string
  music?: any
}

// user types and APIs are handled in api/auth

export async function createDiary(diary: DiaryCreate, userId: number): Promise<DiaryResponse> {
  try {
    const { data } = await api.post<DiaryResponse>(`/diaries?user_id=${userId}`, diary)
    return data
  } catch (err: any) {
    // If 404 (user not found), refresh user_id from username and retry once
    if (err?.response?.status === 404) {
      const username = localStorage.getItem('username')
      if (username) {
        try {
          const user = await getUserInfo(username)
          localStorage.setItem('user_id', String(user.id))
          const { data } = await api.post<DiaryResponse>(`/diaries?user_id=${user.id}`, diary)
          return data
        } catch {}
      }
    }
    throw err
  }
}

export async function getDiaries(userId: number, limit: number = 30): Promise<DiaryResponse[]> {
  try {
    const { data } = await api.get<DiaryResponse[]>(`/diaries?user_id=${userId}&limit=${limit}`)
    return data
  } catch (err: any) {
    if (err?.response?.status === 404) {
      const username = localStorage.getItem('username')
      if (username) {
        try {
          const user = await getUserInfo(username)
          localStorage.setItem('user_id', String(user.id))
          const { data } = await api.get<DiaryResponse[]>(`/diaries?user_id=${user.id}&limit=${limit}`)
          return data
        } catch {}
      }
    }
    throw err
  }
}

export async function getDiary(diaryId: number): Promise<DiaryResponse> {
  const { data } = await api.get<DiaryResponse>(`/diaries/${diaryId}`)
  return data
}
